storm-core/src/jvm/org/apache/storm/topology/StatefulBoltExecutor.java

```diff
@@ -148,4 +148,19 @@ private void fail(List<Tuple> tuples) {
         }
     }
 
+    private static class AckTrackingOutputCollector extends AnchoringOutputCollector {
+        private OutputCollector delegate;
+        private List<Tuple> ackedTuples;
+
+        AckTrackingOutputCollector(OutputCollector delegate) {
+            super(delegate);
+            this.delegate = delegate;
+            this.ackedTuples = new ArrayList<>();
```

Comment:
All access to ackedTuples should ideally be thread safe, as the OutputCollector could be called from a different thread.